feat(settings): persist user management column visibility

pringelmann opened the pull request:
* Resolves: #15632

## Summary

Moves user management column settings (like showing the storage path) from `localStorage` to the database (in user preferences). This ensures configuration persists across browsers, sessions, and devices.

Also refactors the users Vuex store to decouple internal state properties from backend configuration keys, simplifying state management and improving maintainability.

## Checklist

- [x] Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [x] [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [x] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [ ] Screenshots before/after for front-end changes
- [x] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [ ] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)
- [ ] [Labels added](https://github.com/nextcloud/server/labels) where applicable (ex: bug/enhancement, `3. to review`, feature component)
- [ ] [Milestone added](https://github.com/nextcloud/server/milestones) for target branch/version (ex: 32.x for `stable32`)

provokateurin commented on apps/settings/lib/Controller/UsersController.php:
```diff
@@ -250,13 +263,16 @@ public function usersList(INavigationManager $navigationManager, ISubAdmin $subA
 	public function setPreference(string $key, string $value): JSONResponse {
-		$allowed = ['newUser.sendEmail', 'group.sortBy'];
-		if (!in_array($key, $allowed, true)) {
+		$allowedAppValues = ['newUser.sendEmail', 'group.sortBy'];
```
```suggestion
		$allowedKeys = ['newUser.sendEmail', 'group.sortBy'];
```
This would be more precise, no? Alternatively just inline the array, since the variable is only used once.

provokateurin commented on apps/settings/src/store/users.js:
```diff
@@ -47,12 +44,12 @@ const state = {
+		showFirstLogin: usersSettings.showConfig['userList.showFirstLogin'],
+		showLastLogin: usersSettings.showConfig['userList.showLastLogin'],
+		showNewUserForm: usersSettings.showConfig['userList.showNewUserForm'],
+		showLanguages: usersSettings.showConfig['userList.showLanguages'],
```
I wonder if the localStorage values should be migrate, such that you don't lose your current configuration.

provokateurin commented on apps/settings/lib/Controller/UsersController.php:
```diff
@@ -250,13 +263,16 @@ public function usersList(INavigationManager $navigationManager, ISubAdmin $subA
+		if (in_array($key, $allowedAppValues, true)) {
+			$this->config->setAppValue('core', $key, $value);
+		} elseif (in_array($key, self::ALLOWED_USER_PREFERENCES, true)) {
+			$this->config->setUserValue($this->userSession->getUser()->getUID(), $this->appName, $key, $value);
```
These should use IAppConfig and IUserConfig instead.

provokateurin replied:
Looks good, maybe you can get rid of IConfig completely in this class?

provokateurin replied:
Ah yes, for those we still need it.

provokateurin reviewed (changes requested):
The move away from IConfig is somewhat necessary, because it is deprecated usage and we block adding any new code that uses deprecated code. You can check that locally by running psalm.

pringelmann committed:
feat(settings): persist user management column visibility

Previously, column visibility settings were stored in localStorage, causing them to be lost when logging out or switching browsers.
This change moves the persistence to the database as user preferences.
It also refactors the frontend to use clean `userList.*` keys for better consistency between the store and the API.

Signed-off-by: Peter Ringelmann <4850521+Pringels@users.noreply.github.com>

artonge commented on apps/settings/lib/Controller/UsersController.php:
```diff
@@ -59,13 +61,24 @@ class UsersController extends Controller {
+		'user_list_show_last_login',
+		'user_list_show_new_user_form',
+		'user_list_show_languages',
+	];
```
Can we register them in the lexicon?

artonge replied:
Looking at the PR that added it, I think it is an optimisation, even though the constant is not actually used: https://github.com/nextcloud/server/pull/55139

So let's keep it.

pringelmann committed:
refactor(settings): use ConfigLexicon for user list preferences

Signed-off-by: Peter Ringelmann <4850521+Pringels@users.noreply.github.com>

pringelmann committed:
refactor(settings): Re-add user email key to lexicon

Signed-off-by: Peter Ringelmann <4850521+Pringels@users.noreply.github.com>

pringelmann committed:
chore: fix Psalm rebase issue

Signed-off-by: Peter Ringelmann <4850521+Pringels@users.noreply.github.com>

pringelmann committed:
fix(settings): resolve Cypress crash on undefined showConfig

Signed-off-by: Peter Ringelmann <4850521+Pringels@users.noreply.github.com>

Pringels commented:
The other failing cypress test might be flaky since it passes locally and seems unrelated
